rfcs/20200804-primitive-fallback/README.md

```diff
@@ -0,0 +1,172 @@
+constructed having created a valid primitive descriptor implementation.
+
+The options below are listed in the order of increased amount of anticipated
+changes to support primitive fallback.
+
+Additional options are welcome to be added to the list.
+
+### Option 1: leave as is
+
+For this option, the user will take advantage of the existing
+`primitive_desc::next_impl()` API to be able to deploy a different
+implementation if the first one fails. In this case, the user is enforced to
+introduce some control flow, something like
+~~~cpp
+
+auto pd = convolution_forward::primitive_desc(opd, attr, eng);
+convolution_forward prim;
+
+do {
+    try {
+        prim = convolution_forward(pd);
+    }
+    catch (dnnl::error &e) {...}
+} while (prim == convolution_forward() && pd.next_impl());
+~~~
+
+In order to ensure that the issue described above will be taken care of by the
+user, the documentation will be updated accordingly.
+
+Pros: 
+1. Minimal changes required in documentation;
+
+Cons:
+1. The user is responsible for ensuring that a valid primitive implementation
+   exists.
+
+### Option 2: allow construction of an empty primitive
+
+For this option, we attempt to simplify the control flow in [Option
+1](#option-1-leave-as-is) by allowing construction of an empty primitive without
+throwing an exception using the `allow_empty` flag:
+
+~~~cpp
+primitive::primitive(const_dnnl_primitive_desc_t c_pd, bool allow_empty = false)
+~~~
+
+In this case, the control flow can be simplified to:
+
+~~~cpp
+
+auto pd = convolution_forward::primitive_desc(opd, attr, eng);
+convolution_forward prim;
+do {
+    prim = convolution_forward(pd, true);
+} while (prim == convolution_forward() && pd.next_impl());
+
+~~~
+
+Pros:
+1. Minimal changes required in documentation;
+2. A slightly simpler control flow compared to [Option
+   1](#option-1-leave-as-is);
+
+Cons:
+1. The user is responsible for ensuring that a valid primitive implementation exists.
+
+### Option 3: fallback to the reference implementation (personal preference)
+
+For this option we introduce `std::unique_ptr<primitive_desc_t> pd_ref_`,  which
+stores the reference implementation for a particular primitive. in
+`dnnl_primitive_desc_iterator` and in `dnnl_primitive_desc`. In case a JIT
+implementation fails during its generation for any reason, the reference
+implementation will be expected to handle any (or most) use cases supported by
+the library, such as attributes, memory layouts, and data types.
+
+In the context of primitive cache, in order to avoid repeatedly failing to
+generate a JIT primitive implementation, we can:
+1. Set `bool primitive_t::is_initialized_` (which is inherited from `struct
+   c_compatible`) to `false` during construction, and change to `true` if
+   `primitive_t::init(engine_t *engine, bool use_global_scratchpad)` returns
+   `dnnl_success`;
+2. Include `bool primitive_t::is_initialized_` into `primitive_hashing::key_t`
```

Comment:
I'm not sure I understand how the workflow would look, I mean what is the purpose of the new key field?

The problem here is that the value of `bool primitive_t::is_initialized_` is set during primitive creation while the `primitive_hashing::key_t` is created before it. This order is necessary for implementing an efficient multi-threaded cache. 
In other words the key must be known before starting the primitive creation.

It looks like the main reason for a failed JIT generation is large problem size. What if we assume that a failed JIT generation means that the JIT implementation is not applicable for the given problem size. The sizes are part of the key therefore it makes sense to put the "fallback primitive" into the cache to avoid repeatedly failing to
generate a JIT primitive implementation.

Reply:
Indeed @densamoilov , mostly large problem sizes lead to JIT failures currently. If I have understood correctly your suggestion in the last paragraph, we could introduce the same check at `primitive_desc::init` stage, and the implementation will not be even picked up for generation. However, as far as I understood, checking the validity of a primitive using problem sizes hasn't been an option for trivial checks. Plus, it's not really the sizes, but the direct offsets larger than `INT_MAX`. Technically we could circumvent this problem by using indirect loads.

Regarding workflow for primitive cache, thanks for pointing out the way it currently works, I didn't really think it through. The idea was to put the invalid JIT implementation into cache during a `cache_miss`, and at `cache_hit` we check the `is_initialized_` field of the primitive. This is mostly because we have be somehow aware that a JIT implementation will always fail for a given problem size (or other reasons).

Reply:
My point was that it's probably OK to put a primitive containing an invalid JIT implementation into the cache assuming that generation of the JIT implementation always fails.

If we don't want to put such a primitive into the cache then it gets a little trickier because of the promise/future based approach. To do that, we could use `primitive_t::is_initialized_` to invalidate the entry in the same way as we invalidate the entry in case of a runtime error.

Reply:
It looks like that may go with option 1 for this RFC. Resolving the discussion.